### PR TITLE
Enhancement on playlist and features related to watch_later

### PR DESCRIPTION
**Pull request by lhc70000**
**Description:**

- Made it easier for user to resume last played file.
- Show video duration and playback progress in playlist.
  - It calls FFmpeg directly to prefetch video duration for the whole playlist.
- Refresh welcome window when reopened.

deec9fd1-f82b-480a-847f-6ac4d299fe9f
e8c8f216-2653-478a-9f29-ff7b2a350158

TODO

- [ ] Add preference keys

### CODE REVIEW

File: iina/AppDelegate.swift [hunk 1]
```diff
@@ -207,12 +207,16 @@ class AppDelegate: NSObject, NSApplicationDelegate {
     }
   }
 
-  private func showWelcomeWindow() {
+  private func showWelcomeWindow(shouldRefresh: Bool = false) {
```
Comment (saagarjha) [inline iina/AppDelegate.swift]: I'd prefer something along the lines of `showWelcomeWindow(checkingForUpdatedData: Bool = false)`.

File: iina/Extensions.swift [hunk 2]
```diff
@@ -461,6 +461,23 @@ extension NSTextField {
 
 }
 
+extension NSImage {
+  func tinted(withColor tintColor: NSColor) -> NSImage {
```
Comment (saagarjha) [inline iina/Extensions.swift]: `func tinted(_ tintColor: NSColor)`

File: iina/Extensions.swift [hunk 3]
```diff
@@ -461,6 +461,23 @@ extension NSTextField {
...
+    guard self.isTemplate else { return self }
+
+    let image = self.copy() as! NSImage
+    image.lockFocus()
+
+    tintColor.set()
+    NSMakeRect(0, 0, image.size.width, image.size.height).fill(using: .sourceAtop)
```
Comment (saagarjha) [inline iina/Extensions.swift]: Please use `NSRect.init(x:y:width:height:)` instead of `NSMakeRect`.

File: iina/FFmpegController.m [hunk 4]
```diff
@@ -307,4 +307,29 @@ - (void)saveThumbnail:(AVFrame *)pFrame width:(int)width height:(int)height inde
   }
 }
 
++ (double)probeVideoDurationForFile:(nonnull NSString *)file
```
Comment (saagarjha) [inline iina/FFmpegController.m]: Why is this written in C?

File: iina/AppDelegate.swift [hunk 5]
```diff
@@ -207,12 +207,16 @@ class AppDelegate: NSObject, NSApplicationDelegate {
...
     let actionRawValue = Preference.integer(for: .actionAfterLaunch)
     let action: Preference.ActionAfterLaunch = Preference.ActionAfterLaunch(rawValue: actionRawValue) ?? .welcomeWindow
     switch action {
     case .welcomeWindow:
-      PlayerCore.first.initialWindow.showWindow(nil)
+      let window = PlayerCore.first.initialWindow!
+      window.showWindow(nil)
+      if (shouldRefresh) {
```
Comment (alejx) [inline iina/AppDelegate.swift]: Remove the parentheses

File: iina/InitialWindowController.swift [hunk 6]
```diff
@@ -28,8 +29,16 @@ class InitialWindowController: NSWindowController {
...
   @IBOutlet weak var betaIndicatorView: BetaIndicatorView!
+  @IBOutlet weak var lastFileContainerView: InitialWindowViewActionButton!
+  @IBOutlet weak var lastFileIcon: NSImageView!
+  @IBOutlet weak var lastFileNameLabel: NSTextField!
+  @IBOutlet weak var lastPositionLabel: NSTextField!
+  @IBOutlet weak var recentFilesTableTopConstraint: NSLayoutConstraint!
 
-  lazy var recentDocuments: [URL] = NSDocumentController.shared.recentDocumentURLs
+  lazy var recentDocuments: [URL] = {
+    NSDocumentController.shared.recentDocumentURLs.filter { $0 != lastPlaybackURL }
```
Comment (alejx) [inline iina/InitialWindowController.swift]: I think this is weird. Users should be able to see the last file they played in history. They may want to watch that file from very beginning by clicking that file in history.

Comment (lhc70000) [reply iina/InitialWindowController.swift]: It'll resume watch progress regardless of how you open the video. If I don't filter that there will be two completely duplicate records.

File: iina/FFmpegController.m (hunk 4 above)
Comment (lhc70000) [reply iina/FFmpegController.m]: I want to keep all FFmpeg related calls in `FFmpegController`, and Obj-C is a little bit easier to use when interacting with C API.

File: iina/AppDelegate.swift (hunk 5 above)
Comment (lhc70000) [reply iina/AppDelegate.swift]: After-effect of writing too much Javascript

File: iina/InitialWindowController.swift (hunk 6 above)
Comment (ErikMinekus) [reply iina/InitialWindowController.swift]: If "Resume last playback position" is unchecked in Preferences, the files shouldn't resume progress. This change has made it impossible to play the last played file from the beginning.

File: iina/AppDelegate.swift [hunk 7]
```diff
@@ -207,12 +207,16 @@ class AppDelegate: NSObject, NSApplicationDelegate {
     }
   }
 
-  private func showWelcomeWindow() {
+  private func showWelcomeWindow(checkingForUpdatedData: Bool = false) {
```
Comment (ErikMinekus) [inline iina/AppDelegate.swift]: The ```showWelcomeWindow()``` call in ```applicationShouldHandleReopen``` should set this parameter to ```true```.

Comment (alejx) [reply iina/AppDelegate.swift]: Fixed by acc6e511.

**Pull request merged**